### PR TITLE
[SYCLomatic] Refine 4 and support 1 thrust API migration

### PR DESCRIPTION
**Pull request by zhiweij1**
Refined:
thrust::transform
thrust::copy_if
thrust::for_each
thrust::make_zip_iterator
Supported:
thrust::make_reverse_iterator

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/APINamesThrust.inc [hunk 1]
```diff
@@ -1304,3 +1304,115 @@ CONDITIONAL_FACTORY_ENTRY(
...
+          CALL_FACTORY_ENTRY("thrust::copy_if",
+                         CALL(MapNames::getDpctNamespace() + "copy_if",
+                              ARG("oneapi::dpl::execution::seq"),
+                              ARG(0), ARG(1), ARG(2), ARG(3), FUNCTOR(4)))))
+      ))
+)
+
+// thrust::make_zip_iterator
+CALL_FACTORY_ENTRY("thrust::make_zip_iterator",  CALL("oneapi::dpl::make_zip_iterator",  ARG(0)))
+// thrust::make_zip_iterator
```
Comment (zhiweij1) [inline clang/lib/DPCT/APINamesThrust.inc]:
```suggestion
// thrust::make_reverse_iterator
```

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 2]
```diff
@@ -1303,6 +1302,13 @@ makeCallArgCreator(std::string Str) {
   return [=](const CallExpr *C) -> const StringRef { return StringRef(Str); };
 }
 
+std::function<ThrustFunctor(const CallExpr *)>
```
Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp:1305]: add thrust in the name

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 3]
```diff
@@ -1402,14 +1408,44 @@ std::function<std::string(const CallExpr *)>
...
+    E = E->IgnoreImpCasts();
+    if (auto DRE = dyn_cast<DeclRefExpr>(E)) {
+      std::string EnumName = DRE->getNameInfo().getName().getAsString();
+      if (EnumName == "device") {
+        return "oneapi::dpl::execution::make_device_policy(" +
+               makeQueueStr()(C) + ")";
+      } else if (EnumName == "seq" || EnumName == "host") {
+        return "oneapi::dpl::execution::seq";
+      } else {
+        return DRE->getNameInfo().getAsString();
```
Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp]: return enumname

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 4]
```diff
@@ -1402,14 +1408,44 @@ std::function<std::string(const CallExpr *)>
...
-      if (auto DRE = dyn_cast<DeclRefExpr>(ICE->getSubExpr())) {
-        std::string EnumName = DRE->getNameInfo().getName().getAsString();
-        if (EnumName == "device") {
-          return "oneapi::dpl::execution::make_device_policy(" +
-                 makeQueueStr()(C) + ")";
-        } else if (EnumName == "seq") {
-          return "oneapi::dpl::execution::seq";
+    E = E->IgnoreImpCasts();
+    if (auto DRE = dyn_cast<DeclRefExpr>(E)) {
+      std::string EnumName = DRE->getNameInfo().getName().getAsString();
```
Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp:1427]: stringref

**Pull request merged**